content/numerical/NumberTheoreticTransform.h

```diff
@@ -12,43 +12,39 @@
  */
 #pragma once
 
-#include "ModPow.h"
+#include "../number-theory/ModPow.h"
 
 const ll mod = (119 << 23) + 1, root = 3; // = 998244353
 // For p < 2^30 there is also e.g. (5 << 25, 3), (7 << 26, 3),
 // (479 << 21, 3) and (483 << 21, 5). The last two are > 10^9.
 
 typedef vector<ll> vl;
-void ntt(ll* x, ll* temp, ll* roots, int N, int skip) {
-	if (N == 1) return;
-	int n2 = N/2;
-	ntt(x     , temp, roots, n2, skip*2);
-	ntt(x+skip, temp, roots, n2, skip*2);
-	rep(i,0,N) temp[i] = x[i*skip];
-	rep(i,0,n2) {
-		ll s = temp[2*i], t = temp[2*i+1] * roots[skip*i];
-		x[skip*i] = (s + t) % mod; x[skip*(i+n2)] = (s - t) % mod;
-	}
-}
-void ntt(vl& x, bool inv = false) {
-	ll e = modpow(root, (mod-1) / sz(x));
-	if (inv) e = modpow(e, mod-2);
-	vl roots(sz(x), 1), temp = roots;
-	rep(i,1,sz(x)) roots[i] = roots[i-1] * e % mod;
-	ntt(&x[0], &temp[0], &roots[0], sz(x), 1);
+void ntt(vl &a, vl &rt, vl &rev, int n) {
+	rep(i,0,n) if (i < rev[i]) swap(a[i], a[rev[i]]);
+	for (int k = 1; k < n; k *= 2)
+		for (int i = 0; i < n; i += 2 * k)
+			rep(j,0,k) {
+				ll z = rt[j + k] * a[i + j + k] % mod, &ai = a[i + j];
+				a[i + j + k] = z > ai ? ai - z + mod : ai - z;
+				ai = ai + z >= mod ? ai + z - mod : ai + z;
+			}
 }
-vl conv(vl a, vl b) {
-	int s = sz(a) + sz(b) - 1; if (s <= 0) return {};
-	int L = s > 1 ? 32 - __builtin_clz(s - 1) : 0, n = 1 << L;
-	if (s <= 200) { // (factor 10 optimization for |a|,|b| = 10)
-		vl c(s);
-		rep(i,0,sz(a)) rep(j,0,sz(b))
-			c[i + j] = (c[i + j] + a[i] * b[j]) % mod;
-		return c;
+
+vl conv(const vl &a, const vl &b) {
+	if (a.empty() || b.empty())
+		return {};
+	int s = sz(a) + sz(b) - 1, L = 32 - __builtin_clz(s), n = 1 << L;
+	vl inl(a), inr(b), out(n), rt(n, 1), rev(n);
+	inl.resize(n), inr.resize(n);
+	rep(i,0,n) rev[i] = (rev[i / 2] | (i & 1) << L) / 2;
+	int curL = mod / 2;
+	for (int k = 2; k < n; k *= 2) {
+		ll z[] = {1, modpow(root, curL /= 2)};
+		rep(i,k,2*k) rt[i] = rt[i / 2]* z[i & 1]%mod;
 	}
-	a.resize(n); ntt(a);
-	b.resize(n); ntt(b);
-	vl c(n); ll d = modpow(n, mod-2);
-	rep(i,0,n) c[i] = a[i] * b[i] % mod * d % mod;
-	ntt(c, true); c.resize(s); return c;
+	ntt(inl, rt, rev, n), ntt(inr, rt, rev, n);
+	ll invN = modpow(n, mod - 2);
+	rep(i,0,n) out[-i&(n-1)] = (inl[i] * inr[i] % mod)* invN%mod;
```

Comment:
no need for the parentheses. Space before/after %; if that causes the line to exceed 63 chars, you can rename invN to inv

Speaking of renames, we could also rename inl/inr to e.g. L/R, which is harder to typo

Reply:
I don't like capital variable names :sweat_smile: although your rationale makes sense.

I changed the spacing and renamed `invN` to `inv`. If you think it's worth it to change it to `L/R`, can you do it?

Reply:
Not seeing this change. I can make the L/R change; I think it's worth it. (I somewhat like capital letter names, actually - it makes them stand out, and they aren't that much harder to type. I almost always use them for problem inputs.)

Reply:
I also use them for problem inputs, but *only* problem inputs. So it feels weird to me to use them elsewhere. Amended the commit with the actual change...

Reply:
Also made the `L`/`R` change since you're set on it.